ipfs-cluster/trigger-server/triggerserver.go

```diff
@@ -8,88 +8,133 @@ import (
     "fmt"
     "io/ioutil"
     "math/rand"
-    "strconv"
     "os"
     "github.com/ipfs/go-ipfs-api"
+    "encoding/json"
+    "io"
     "bytes"
 )
 
-const configPath = "/etc/trigger-server"
-// port is HTTP API listen port
-const port = 8082
-var runnerPath = os.Getenv("GOPATH") + "src/github.com/ipfs/kubernetes-ipfs/ipfs-cluster/"
-// Runner takes two args
-const runnerNumNodes = "5"
-const runnerNumPins = "5"
+// TriggerConfig is server configuration JSON format
+type TriggerConfig struct {
+	Secret string
+	Port int
+	RunnerPath string
+	Runner string
+	RunnerArgs string
+}
 
-// Secret is set by main
-var secret = ""
+const configPath = "/etc/trigger-server.json"
+var config TriggerConfig
 
 // IPFS Shell
 var sh = shell.NewShell("localhost:5001")
 
 func run(rw http.ResponseWriter, req *http.Request) {
-    if req.Method == "POST" {
-	if req.Header.Get("X-Auth") != secret {
-		log.Println("User failed authentication")
-		fmt.Fprintf(rw, "Invalid authentication\n")
-	} else {
-		// This command takes a while, so we'll hang out here while it runs
-		cmd := exec.Command("bash", "runner.sh", runnerNumNodes, runnerNumPins)
-		cmd.Dir = runnerPath
-		out, err := cmd.CombinedOutput()
-		// Add the output to IPFS
-		addr, ipfsErr := sh.Add(bytes.NewReader(out))
-		// Make it an IPFS link
-		addr = "/ipfs/" + addr
-		// Send the link to the client
-		fmt.Fprintf(rw, addr)
-		// Handle errors that may have occurred
+    if req.Method != "POST" {
+        fmt.Fprintf(rw, `Kubernetes-IPFS Trigger Server
+	https://github.com/ipfs/kubernetes-ipfs/
+	`)
+	return
+    }
+    if req.Header.Get("X-Auth") != config.Secret {
+	    log.Println("User failed authentication")
+	    rw.WriteHeader(http.StatusUnauthorized)
+	    fmt.Fprintf(rw, "Invalid authentication\n")
+	    return
+    }
+    // This command takes a while, so we'll hang out here while it runs
+    cmd := exec.Command("bash", "-c", fmt.Sprintf("%s%s %s", config.RunnerPath, config.Runner, config.RunnerArgs))
+    cmd.Dir = config.RunnerPath
+    pipeReader, pipeWriter := io.Pipe()
```

Comment:
Perhaps implement a custom io.Writer. The `Write` method in it just writes the bytes to both the response writer and your `out` buffer. That way you don't need a go-routine on the side reading from one reader and writing somewhere else.